### ISSUE TITLE
Split Book entity

### ISSUE THREAD

**Issue by knjk04** [Project-Books/book-project#198]
**Context**

The `Book` entity currently includes the following fields:

>  // For books that have been read
    private RatingScale rating;
    private LocalDate dateStartedReading;
    private LocalDate dateFinishedReading;
    private String bookReview;

The comment indicates that those four fields are only applicable to books that have been read, and not just any book (e.g. a book that is yet to be read).

**What needs to be done**

The `Book` entity could be split into a new class that inherits from `Book` called `ReadBook` with those fields.

**Comment by edwardUL99** [comment]
I can take a look at this if you'd like?

**Comment by edwardUL99** [comment]
The getEdition method seems to be missing, should I add that too?

**Comment by knjk04** [comment]
> I can take a look at this if you'd like? Just let me know for the four fields which are required to be known at construction or any getters/setters for these

That would be great, thanks! The constructor for this new class should take the same arguments as the explicit constructor in the `Book` class (title and author).

**Comment by knjk04** [comment]
> The getEdition method seems to be missing, should I add that too?

Lombok should provide a getter for that as we're using the `@Data` annotation (see https://projectlombok.org/features/Data)

> I've noticed, particularly in BookForm there are method calls to getDateStartedReading but if we move that to a different class those methods no longer exist in Book. I saw there's an issue for a form for a review on a read shelf, so should these be removed from BookForm?

This is a good point, I was just been thinking about this. `BookForm` will need to be refactored. There's a couple of ways this could be done, but I think it would be better if we didn't remove any fields from being bound in `BookForm`. One way would be to use a `Factory`. 

Let me know if you need anything else

**Comment by edwardUL99** [comment]
What should the factory do?

**Comment by knjk04** [comment]
The factory pattern can be used to decide which subclass to construct.

You could also make `Book` a mapped superclass and make entity subclasses for `ReadBook`, `ToReadBook`, etc.

> I'm thinking that as of issue #171 , a normal book in the BookShelf shouldn't have fields for review or rating if they haven't been read right? These fields could be added for the case of it being on the read shelf

Yes, only books that are in the 'read' shelf should have the option of a user setting the rating and the review. I don't know whether I've misunderstood you here, but we're hiding the rating field in the `BookForm` if the shelf chosen was not `read` (as per issue #171 ).

As you correctly pointed out, the rating and review fields (amongst others) are only applicable to `read` books, which is why this issue wants to split the `Book` class.

**Comment by edwardUL99** [comment]
Perfect I'll try and take a look at that. It might take me a while because I'm currently working on another project and work too. So if it needs to be done soon feel free to assign it to someone else

**Comment by knjk04** [comment]
Do you know roughly how long it'll take you?

**Comment by edwardUL99** [comment]
@knjk04 I'll take a look at it tonight and see how long I think it will take me. I actually just got almost finished up with that other project just there so I should have a bit extra time now. Don't think it should take too long as it's only one method in BookForm that has errors

**Comment by edwardUL99** [comment]
You can unassign me from it. Sorry about this but I'll contribute to something when I'm somewhat less busy👍

**Comment by MarQs1990** [comment]
Hi, I would have a look at this. I will be in hospital next week, so I have enough time to rearrange the book entities.
I think there are 3 types of books
1. ToReadBook (open for reading)
2. ReadingBook (reading in progress)
3. ReadBook (done reading)

I don't know if it is sensefull, to develop 3 different classes. Maybe we add a builder pattern and have the 2 mandatory fields by building the book and by switching the state of the book the other fields can (or maybe must) be set.
I'm thinking of, that a book can be added in all three states. After adding the book the user can switch the state of the book and by doing that other fields can be set (e.g. switching from open to in progress the user will set the start date)

**Comment by knjk04** [comment]
@edwardUL99 No matter

**Comment by knjk04** [comment]
@MarQs1990 Thanks, I'll assign you to the issue. 

That sounds good. Another type of book is a `DidNotFinishBook`.

I look forward to seeing how you implement it.

**Comment by knjk04** [comment]
@MarQs1990 We haven't heard back from you in a while, so I just wanted to follow up. How are you getting on with this issue? If you need a hand, please let us know over Gitter.

This is a friendly check-up message, as per our contributing document. If we do not hear back from you within seven days, we will unassign you from this issue so that someone else can pick this up.

**Comment by MarQs1990** [comment]
@knjk04 as i said i will look at this, when i'm in hospital. now the time has come :-) i will work on it today

**Comment by knjk04** [comment]
@MarQs1990 Apologies, I forgot about that. Hope you are OK.

**Comment by MarQs1990** [comment]
@knjk04 I have looked over the code and made some thoughts of this issue.
Why do you need more book entities?
I mean the user can add a book with different parameters and the parameters are checked for a valid value and even if they're not filled, they are added to the newly created book object.
So what for do you need more book entities? What do you think is getting better or easier with the a ToReadBook and FinishedReadingBook class?

**Comment by knjk04** [comment]
We are currently relying on our frontend to only take in fields that correspond to a given book type. However, it would be nice if this was in harmony with our backend. 

We don't need necessarily need more book entities, but it would be nice to have a way to separate `Book` into different abstractions. As it stands, the `Book` entity is only going to get larger as we are planning to add more fields to fit. It doesn't make much sense to me if only some of those fields pertain to a, for example, read book. 

This all goes back to my first comment where we would have rather some kind of abstraction on the backend so that fields that only pertain to a particular kind of book can only be added to that type of book.

I believe the readability of the code will be improved by making classes like `ToReadBook`. When writing tests or populating fields in `PredefinedShelfService` with some test data, it would be nicer if a developer knows by the type alone which fields are accessible and which are not, to prevent some fields being unnecessarily filled in (e.g. the rating field for a 'to read' book). In addition, maintainability can be improved.

**Comment by knjk04** [comment]
@MarQs1990 As we have not heard from you in a while, I am going to unassign you from the issue so that someone else can pick this up (as per our contributing document)

**Comment by GabrielCsik** [comment]
Hello @knjk04  I'm new here and I would like to contribute to the project. Can I work on this issue?

**Comment by Hassan-Elseoudy** [comment]
I'll take this if possible? @knjk04

**Comment by knjk04** [comment]
@Hassan-Elseoudy Thanks for your interest in this! Before I assign you, could you please describe your implementation approach?

**Comment by Hassan-Elseoudy** [comment]
I'm thinking of applying State design pattern by creating more classes `states` , in this case `Entities`, A User's book can be only In one state, So for the database, when moving one state to another, a record will be deleted. and a record will be inserted in the new state table.

Example:

Book A begin with state `Not Started` then I perform action `moveToNextState()`, it moves to `In Progress`, (Now the book is deleted from user's `Not Started` list and in `In Progress` state) then to `Done`.

A book shouldn't be aware of page read, started and such Attributes.
A user will only be considered about the attribute of book at its state.

Cheers.

**Comment by knjk04** [comment]
@Hassan-Elseoudy Sounds good, I'll assign you to the issue. Thanks!

**Comment by knjk04** [comment]
@Hassan-Elseoudy I've removed the 'hacktoberfest' label as I don't think we'll be able to review & merge this in before the end of the month. Are you still happy to work on this?

**Comment by Hassan-Elseoudy** [comment]
> @Hassan-Elseoudy I've removed the 'hacktoberfest' label as I don't think we'll be able to review & merge this in before the end of the month since I still believe it needs to be rewritten. Are you still happy to work on this even without the 'hacktoberfest' label?

I'm, yes

**Comment by cw40** [comment]
I know some other people are working on this, but I'd love to take a look at this too.

**Comment by knjk04** [comment]
@cw40 Thanks for your interest in this. This is currently blocked, but I can get back to you when we decide to unblock this
